pkg/cmd/roachtest/tests/loss_of_quorum_recovery.go

```diff
@@ -0,0 +1,291 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package tests
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/util/contextutil"
+	"github.com/cockroachdb/cockroach/pkg/workload/histogram"
+)
+
+func registerLOQRecovery(r registry.Registry) {
+	spec := r.MakeClusterSpec(6)
+	r.Add(registry.TestSpec{
+		Name:              "loqrecovery/upreplicate=true",
+		Owner:             registry.OwnerKV,
+		Tags:              []string{`default`},
+		Cluster:           spec,
+		NonReleaseBlocker: true,
+		Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
+			recoverLossOfQuorum(ctx, t, c, true)
+		},
+	})
+	r.Add(registry.TestSpec{
+		Name:              "loqrecovery/upreplicate=false",
+		Owner:             registry.OwnerKV,
+		Tags:              []string{`default`},
+		Cluster:           spec,
+		NonReleaseBlocker: true,
+		Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
+			recoverLossOfQuorum(ctx, t, c, false)
+		},
+	})
+}
+
+type testOutcomeMetric int
+
+const (
+	// Nodes fail to start after recovery was performed.
+	restartFailed testOutcomeMetric = iota
+	// Nodes restarted but decommission didn't work.
+	decommissionFailed
+	// Nodes started, but workload failed to succeed on any writes after restart.
+	workloadFailed
+	// Nodes restarted and workload produced some writes.
+	success
+)
+
+func recoverLossOfQuorum(
+	ctx context.Context, t test.Test, c cluster.Cluster, waitForUpreplicate bool,
+) {
+	nodes := c.Spec().NodeCount - 1
+	controller := c.Spec().NodeCount
+	c.Put(ctx, t.Cockroach(), "./cockroach", c.All())
+	startOpts := option.DefaultStartOpts()
+	settings := install.MakeClusterSettings()
+	c.Start(ctx, t.L(), startOpts, settings, c.Range(1, nodes))
+
+	db := c.Conn(ctx, t.L(), 1)
+	defer db.Close()
+
+	t.L().Printf("Started cluster")
+
+	// Lower the dead node detection threshold to make decommissioning nodes
+	// faster.
+	_, err := db.Exec("SET CLUSTER SETTING server.time_until_store_dead = '1m30s'")
+	if err != nil {
+		t.Fatalf("Failed to set dead timeout to 1m: %s", err)
```

Comment:
nit: this doesn't match the 1m30s above. I'd just drop the duration from the message.

Also, why don't we set it any lower?

Reply:
1:30 is the lowest allowed value actually. I wanted to set it to something very low as I only expect failure to be deliberate.